### DISCUSSION TITLE
Unable to apply web build patch 2.19.0

### DISCUSSION THREAD

**Discussion by hirodragon** [dani-garcia/vaultwarden#1524] [Q&A] [answered]
### Subject of the issue
Unable to apply web build patch 2.19.0 from https://github.com/dani-garcia/bw_web_builds/tree/master/patches
### Deployment environment
Rasbian 10 (Buster)

* bitwarden_rs version:
1.19.0

* Install method:
Building own binary

### Steps to reproduce

`git clone https://github.com/bitwarden/web.git web-vault.git`
`cd web-vault.git`
`git checkout v2.19.0`
`curl -O https://raw.githubusercontent.com/dani-garcia/bw_web_builds/master/patches/v2.19.0.patch`
`git apply v2.19.0.patch`

### Expected behaviour

Patch is applied.

### Actual behaviour

Patch is not applied, error message occurs:
`error: jslib/src/angular/components/register.component.ts: No such file or directory`
`error: patch failed: src/app/services/services.module.ts:135`
`error: src/app/services/services.module.ts: patch does not apply`
Same behaviour for 2.18.1 patch, last patch succesfully applied was 2.18.0.

**Comment by Nyxtorm** [comment]
Hello,

Your jslib folder should be empty. You should make a clone of the repository https://github.com/bitwarden/jslib.git and copy its contents to your empty web-vault's jslib folder.

For example : 

```
git clone https://github.com/bitwarden/web.git web-vault.git
git clone https://github.com/bitwarden/jslib.git
rsync -rtlDvP  jslib.git/ web-vault.git/jslib/
cd web-vault.git
git checkout v2.19.0
curl -O https://raw.githubusercontent.com/dani-garcia/bw_web_builds/master/patches/v2.19.0.patch
git apply v2.19.0.patch
```

Before, the jslib folder was included in the web-vault repository. He now has its own repository.

Sorry for my English and hoping to have helped.

EDIT : small correction in the rsync paths, but you can use any other copy tool (cp, etc ...)

**Comment by Nyxtorm** [comment]
I suggest a better solution : 

`git clone --recurse-submodules https://github.com/bitwarden/web.git web-vault.git`

instead of

`git clone https://github.com/bitwarden/web.git web-vault.git`

This way, just one clone is enough, git will automatically get the submodules (so the jslib repository in this case)

EDIT : Moreover, the documentation is up to date on this subject: https://github.com/dani-garcia/bitwarden_rs/wiki/Building-binary#install-the-web-vault 😉

**Comment by BlackDex** [comment, marked as answer]
Or you can run the following command after a simple clone within the cloned folder:
```bash
git submodule update --init --recursive
```

**Reply by hirodragon** [reply]
It fixed my problem, thank you.

**Reply by dinger1986** [reply]
Thanks, working for me, scripts updated

**Reply by twiclo** [reply]
I'm running into this same problem now with v2.24.0
```[twiclo@toph web-vault]$ git apply v2.24.0.patch
error: patch failed: jslib/angular/src/components/register.component.ts:41
error: jslib/angular/src/components/register.component.ts: patch does not apply
error: patch failed: src/app/organizations/settings/organization-subscription.component.ts:112
error: src/app/organizations/settings/organization-subscription.component.ts: patch does not apply
error: patch failed: src/app/organizations/vault/vault.component.ts:62
error: src/app/organizations/vault/vault.component.ts: patch does not apply
error: patch failed: src/app/vault/vault.component.ts:77
error: src/app/vault/vault.component.ts: patch does not apply
```

This was after running `git submodule update --init --recursive` as suggested in the "Building binary" guide

**Reply by BlackDex** [reply]
Did you also checked out the extract same commit? Else the patch will not work

**Reply by twiclo** [reply]
```
[twiclo@toph web-vault]$ git checkout "v2.24.0"
M	jslib
Previous HEAD position was d10dc94a Remove old 'release' ref in workflow (#1328)
HEAD is now at 9e3d45da changing how we checkout the new release branch
[twiclo@toph web-vault]$ git apply v2.24.0.patch
error: patch failed: jslib/angular/src/components/register.component.ts:41
error: jslib/angular/src/components/register.component.ts: patch does not apply
```

**Reply by twiclo** [reply]
Figured it out. I had to re-run the git submodule command after checking out to the right tag
